server/webapp/WEB-INF/rails/webpack/views/pages/new_agents/environment_dropdown_button.tsx

```diff
@@ -61,4 +71,88 @@ export class EnvironmentsDropdownButton extends AbstractDropdownButton<EnvAttrs>
   protected hasAssociationWith(agent: Agent, environment: string) {
     return agent.environmentNames().includes(environment);
   }
+
+  protected buildTriStateCheckBox(vnode: m.Vnode<DropdownAttrs & EnvAttrs>) {
+    this.triStateCheckboxMap.clear();
+
+    const selectedAgentsUUID = vnode.attrs.agentsVM.selectedAgentsUUID();
+    const allAgents = vnode.attrs.agentsVM.all();
+
+    const allSelectedAgents = allAgents.filter((agent) => selectedAgentsUUID.includes(agent.uuid));
+    const allSelectedAgentsEnvs = allSelectedAgents.map((agent) => agent.environments);
+    const allAgentsEnvs = _.flatMap(allSelectedAgentsEnvs).map((env) => env.name);
+
+    const allUniqueEnvs = _.uniq(allAgentsEnvs.concat(this.data())).sort();
+
+    allUniqueEnvs.map((item) => {
+      const numberOfAgentsWithSpecifiedEnv = vnode.attrs.agentsVM.list().reduce((count: number, agent: Agent) => {
+        if (vnode.attrs.agentsVM.isAgentSelected(agent.uuid) && this.hasAssociationWith(agent, item)) {
+          count += 1;
+        }
+        return count;
+      }, 0);
+
+      const disabledDueToConfigRepoAssociation = EnvironmentsDropdownButton.isAssociatedViaConfigRepo(item, allSelectedAgents);
+      const disabledDueToUnknown = EnvironmentsDropdownButton.isUnknown(item, allSelectedAgents);
+
+      const disabled = disabledDueToConfigRepoAssociation || disabledDueToUnknown;
+      const msg = EnvironmentsDropdownButton.getTooltipContent(disabledDueToConfigRepoAssociation, disabledDueToUnknown);
+
+      switch (numberOfAgentsWithSpecifiedEnv) {
+        case 0:
+          this.triStateCheckboxMap.set(item, new EnvironmentInfo(TristateState.off, disabled, msg));
+          break;
+        case vnode.attrs.agentsVM.selectedAgentsUUID().length:
+          this.triStateCheckboxMap.set(item, new EnvironmentInfo(TristateState.on, disabled, msg));
+          break;
+        default:
+          this.triStateCheckboxMap.set(item, new EnvironmentInfo(TristateState.indeterminate, disabled, msg));
+          break;
+      }
+    });
+  }
+
+  private static isAssociatedViaConfigRepo(envName: string, agents: Agent[]): boolean {
```

Comment:
No need to do this over here. Can we move it to `EnvironmentInfo`?

Reply:
Refactored the method in question. Moved all the logic pertaining to envs to a single loop.